### DISCUSSION TITLE
Adding images into a shape while mapping through data

### DISCUSSION THREAD

**Discussion by hyfydistro** [airbnb/visx#1307] [Q&A] [answered]
Is there a way to add/embed/clip images into a `<circle>` (lowercase "c") in the visx <Pack/> library?  

I'm looking to make a chart like the following. I want to surround the main person by other bubble (people) whom have different values (which should explain why the bubble is larger).
![graph-05](https://user-images.githubusercontent.com/24542308/128304469-8de28d6e-a21f-4656-a9c0-5150ce743baa.png)

**Comment by williaster** [comment, marked as answer]
Hey @hyfydistro 👋  thanks for checking out `visx`. I think to do this you'd just need to lean on the `SVG` api for using an image as a fill. One way to do this is to create a `<pattern />`  for each image within a `<defs />` tag, then use those patterns as the fill for the circle. Note that pattern id's that you define should be unique on the page else you will have overlap/non-determinant behavior. 

Here's a [sandbox demo](https://codesandbox.io/s/vigorous-jang-999gh?file=/Example.tsx), hope it helps!

<img src="https://user-images.githubusercontent.com/4496521/128428539-51aa37cb-4b5f-4e8e-bbcf-379da8804e2d.png" width="400" />

**Reply by williaster** [reply]
based on your question about non-overlapping circles: sorry I assumed you already had x/y/radius coming from somewhere. if you don't you need to use an algorithm to layout the circles so they don't overlap.

there are many approaches to that but one which is easy in `visx`/`d3` is circle packing for which we have a demo [here](https://airbnb.io/visx/pack) (have you looked at the gallery?).

<img src="https://user-images.githubusercontent.com/4496521/128553577-c93a3dfc-0044-41cd-a160-8a66603db589.png" width="400" />

I didn't account for labels here, and you can remove the outermost circle (I left it to demonstrate what is going on) but here's an [updated sandbox](https://codesandbox.io/s/mystifying-panini-l5iml?file=/Example.tsx).

<img src="https://user-images.githubusercontent.com/4496521/128555392-590b1641-a3b8-4c1e-8f0c-8951ffdc0b48.png" width="400" />

**Reply by hyfydistro** [reply]
Thanks! I was able to get it working nicely!

**Reply by hyfydistro** [reply]
Could you recommend me what method I should use to add tooltip on each bubble? @williaster I am planning add an "on-hover" information but I am lost at which options to choose. In addition, I want to add an "on-click" event that will get the data information of the bubble. I could handle that with state management, but is it possible with a tooltip wrapper?

**Reply by williaster** [reply]
Hi @hyfydistro I would check out the `@visx/tooltip` demo for triggering and rendering tooltips https://airbnb.io/visx/tooltip

**Reply by hyfydistro** [reply]
Thanks! I was able to get it in with `useTooltipInPortal`. Not sure why it didn't work with the basic `Tooltip`. Is there a difference or situational method?

**Comment by hyfydistro** [comment]


**Reply by hyfydistro** [reply]
All good! I just had to restart server! ^^;

**Comment by hyfydistro** [comment]
Is it possible for the bubbles not to overlap each other? I thought the <Pack /> had something but I'm having difficulty. @williaster
